tapa/common/floorplan/gen_slot_cpp.py

```diff
@@ -0,0 +1,187 @@
+"""Generate floorplan slot cpp for hls synth."""
+
+__copyright__ = """
+Copyright (c) 2024 RapidStream Design Automation, Inc. and contributors.
+All rights reserved. The contributor(s) of this file has/have agreed to the
+RapidStream Contributor License Agreement.
+"""
+
+import pathlib
+import re
+from string import Template
+
+_SCALAR_PRAGMA = (
+    "#pragma HLS interface ap_none port = {name} register\n"
+    "  {{ auto val = reinterpret_cast<volatile uint8_t &>({name}); }}"
+)
+_FIFO_IN_PRAGMA = (
+    "#pragma HLS disaggregate variable = {name}\n"
+    "#pragma HLS interface ap_fifo port = {name}._\n"
+    "#pragma HLS aggregate variable = {name}._ bit\n"
+    "  void({name}._.empty());\n"
+    "  {{ auto val = {name}.read(); }}\n"
+)
+_FIFO_OUT_PRAGMA = (
+    "#pragma HLS disaggregate variable = {name}\n"
+    "#pragma HLS interface ap_fifo port = {name}._\n"
+    "#pragma HLS aggregate variable = {name}._ bit\n"
+    "  void({name}._.full());\n"
+    "  {name}.write({type}());"
+)
+
+_PRAGMA = {
+    "scalar": _SCALAR_PRAGMA,
+    "mmap": _SCALAR_PRAGMA,
+    "istream": _FIFO_IN_PRAGMA,
+    "ostream": _FIFO_OUT_PRAGMA,
+    "istreams": _FIFO_IN_PRAGMA,
+    "ostreams": _FIFO_OUT_PRAGMA,
+}
+
+_SCALAR_PORT_TEMPLATE = "{type} {name}"
+_PORT_TEMPLATE = {
+    "istream": "tapa::istream<{type}>& {name}",
+    "ostream": "tapa::ostream<{type}>& {name}",
+    "istreams": "tapa::istream<{type}>& {name}",
+    "ostreams": "tapa::ostream<{type}>& {name}",
+    "scalar": _SCALAR_PORT_TEMPLATE,
+    "mmap": _SCALAR_PORT_TEMPLATE,
+    "async_mmap": _SCALAR_PORT_TEMPLATE,
+}
+_DEF_TEMPLATE_PATH = pathlib.Path(__file__).parent / "template/slot_def_template.txt"
+_DECL_TEMPLATE_PATH = pathlib.Path(__file__).parent / "template/slot_decl_template.txt"
+
+
+def gen_slot_cpp(slot_name: str, top_name: str, ports: list, top_cpp: str) -> str:
+    """Generate floorplan slot cpp for hls synth.
+
+    slot_name: Name of the slot
+    ports: List of ports in the slot. Each port should match port format
+        in tapa graph dict.
+        e.g.
+        {
+            "cat": "istream",
+            "name": "a",
+            "type": "float",
+            "width": 32
+        }
+    """
+
+    cpp_ports = []
+    cpp_pragmas = []
+    for port in ports:
+        assert isinstance(port, dict)
+        assert "cat" in port
+        assert "name" in port
+        assert "type" in port
+        assert "width" in port
+        assert port["cat"] in _PORT_TEMPLATE
+        port_type = port["type"]
+        port_cat = port["cat"]
+
+        # if "name[idx]" exists, replace it with "name_idx"
+        name = re.sub(r"\[([^\]]+)\]$", lambda m: f"_{m.group(1)}", port["name"])
+
```

Comment:
`idx` is not necessarily a int literal; it can also be a constexpr variable. Do we want to at least detect this and err?